api.go

```diff
@@ -0,0 +1,218 @@
+		ConnectStart: func(_, _ string) {
+			if t1.IsZero() {
+				// connecting to IP
+				t1 = time.Now()
+			}
+		},
+		ConnectDone: func(net, addr string, err error) {
+			if err != nil {
+				log.Printf("unable to connect to host %v: %v", addr, err)
+			}
+			t2 = time.Now()
+		},
+		GotConn:              func(_ httptrace.GotConnInfo) { t3 = time.Now() },
+		GotFirstResponseByte: func() { t4 = time.Now() },
+	}
+
+	// We'll use our Context in this request in case we have to shut down midstream
+	req = req.WithContext(httptrace.WithClientTrace(ctx, trace))
+
+	resp, err := client.Do(req)
+	if err != nil {
+		log.Println("Failed to read response:", err)
+		return
+	}
+
+	// Send our server response code as an event
+	storage.EventDistributor <- j.makeEvent(resp.StatusCode)
+
+	// unmarshal response and save into map
+	if responses[j.Name], err = ioutil.ReadAll(resp.Body); err != nil {
+		log.Println("WARNING!", err)
+	} else {
+		out, _ := replacePlaceholders(fmt.Sprintf("%s response:\n{{ %s }}\n", j.Name, j.Name), responses)
+		log.Println(out)
+	}
+
+	// Even though we never read the response body, if we don't close it,
+	// the http.Transport goroutines will terminate and the app will eventually
+	// crash due to OOM
+	resp.Body.Close()
+
+	t5 := time.Now() // after read body
+	if t0.IsZero() {
+		// we skipped DNS
+		t0 = t1
+	}
+
+	url, err := url.Parse(j.URL)
+	if err != nil {
+		log.Println("Failed to parse URL:", err)
+		return
+	}
+
+	switch url.Scheme {
+	case "https":
+		storage.MetricDistributor <- j.makeMetric("dns_duration_milliseconds", t1.Sub(t0).Seconds()*1000)
+		storage.MetricDistributor <- j.makeMetric("server_connection_duration_milliseconds", t2.Sub(t1).Seconds()*1000)
+		storage.MetricDistributor <- j.makeMetric("tls_handshake_duration_milliseconds", t3.Sub(t2).Seconds()*1000)
+		storage.MetricDistributor <- j.makeMetric("server_processing_duration_milliseconds", t4.Sub(t3).Seconds()*1000)
+		storage.MetricDistributor <- j.makeMetric("server_response_duration_milliseconds", t5.Sub(t4).Seconds()*1000)
+		storage.MetricDistributor <- j.makeMetric("time_to_first_byte_milliseconds", t4.Sub(t0).Seconds()*1000)
+
+	case "http":
+		storage.MetricDistributor <- j.makeMetric("dns_duration_milliseconds", t1.Sub(t0).Seconds()*1000)
+		storage.MetricDistributor <- j.makeMetric("server_connection_duration_milliseconds", t3.Sub(t1).Seconds()*1000)
+		storage.MetricDistributor <- j.makeMetric("server_processing_duration_milliseconds", t4.Sub(t3).Seconds()*1000)
+		storage.MetricDistributor <- j.makeMetric("server_response_duration_milliseconds", t5.Sub(t4).Seconds()*1000)
+		storage.MetricDistributor <- j.makeMetric("time_to_first_byte_milliseconds", t4.Sub(t0).Seconds()*1000)
+	}
+}
+
+
+func addHeaders(req *http.Request, j JobStep, responses map[string]json.RawMessage) error {
+	req.Header = http.Header{}
+
+	for k, value := range j.Header {
+		key := textproto.CanonicalMIMEHeaderKey(k)
+		if _, ok := req.Header[key]; !ok {
```

Comment:
I think any duplicate headers should be overwritten instead of appended to the existing header

Reply:
I was thinking about the situation where we have `x-header-tag` and `X-Header-Tag` with different values. We don't know which will come first in the map or which the user wants so I think it makes sense to leave both in this case.

Reply:
x-header-tag and X-Header-Tag are supposed to be the same header according to spec IIRC. I see your point though, but taking a look at the source code for [header.go](https://golang.org/src/net/http/header.go?s=1710:1753#L27) seems like just using `req.Header.Add` would work here.